// samsung-tv-control/lib/auto-search.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const node_ssdp_1 = require("node-ssdp");
const request = require("request");
const SAMSUNG_TV_URN = 'urn:samsung.com:device';
class AutoSearch {
    constructor() {
        this.IPs = [];
        this.TVs = [];
        this.client = new node_ssdp_1.Client();
        this.client.on('response', this.deviceUpdate.bind(this));
    }
    search(time) {
        return new Promise((resolve, reject) => {
            this.client.search('ssdp:all');
            setTimeout(this.stopSearch.bind(this, resolve, reject), time || 5000);
        });
    }
    deviceUpdate(headers, _, rinfo) {
        if (headers && headers.ST && headers.ST.includes(SAMSUNG_TV_URN) && !this.IPs.includes(rinfo.address)) {
            this.IPs.push(rinfo.address);
            request.get({ url: `http://${rinfo.address}:8001/api/v2/` }, (err, res) => {
                if (!err && res.statusCode === 200) {
                    const data = JSON.parse(res.body);
                    this.TVs.push({
                        name: data.device.name,
                        model: data.device.modelName,
                        ip: data.device.ip,
                        mac: data.device.wifiMac
                    });
                }
            });
        }
    }
    stopSearch(resolve) {
        this.client.stop();
        resolve(this.TVs);
    }
}
exports.default = AutoSearch;
